File: c++/142. Linked List Cycle II.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *detectCycle(ListNode *head) {
        unordered_set<ListNode*> us;
        while (head) {
            if (us.find(head) == us.end())
                us.insert(head);
            else
                return head;
            head = head->next;
        }
        return head;
    }
};

class Solution {
public:
	ListNode * detectCycle(ListNode *head) {
		if (head == NULL || head->next == NULL) return NULL;
		ListNode* slow = head->next, *fast = head->next->next;
		while (fast && fast->next) {
			if (slow == fast)
				break;
			slow = slow->next;
			fast = fast->next->next;
		}

		fast = head;
		while (slow != fast) {
			slow = slow->next;
			fast = fast->next;
		}

		return fast;
	}
};

File: python/142. Linked List Cycle II.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution(object):
    def detectCycle(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if not head or not head.next: return None
        f = False
        slow = head.next
        fast = head.next.next
        while fast and fast.next:
            if slow==fast:
                f = True
                break
            slow = slow.next
            fast = fast.next.next
        if not f: return None

        fast = head
        while fast != slow:
            slow = slow.next
            fast = fast.next
        return fast

if __name__ == '__main__':
    head = ListNode(1)
    head.next = ListNode(2)
    su = Solution()
    node = su.detectCycle(head)                    
    print node